src/mesh/duplicate_nodes.cc

```diff
@@ -0,0 +1,256 @@
+{
+  // initialize the objects from edges
+  FOR_EDGES(mesh_, edge) {
+    switch (edge->side(0)->dim()) {
+      case 0:
+        {
+          MeshObject<0> p;
+          p.nodes[0] = obj_4_node_[mesh_->node_vector.index(edge->side(0)->node(0))];
+          obj_4_edg_.push_back(points_.size());
+          points_.push_back(p);
+        }
+        break;
+      case 1:
+        {
+          MeshObject<1> l;
+          for (unsigned int i=0; i<2; i++)
+            l.nodes[i] = obj_4_node_[mesh_->node_vector.index(edge->side(0)->node(i))];
+          obj_4_edg_.push_back(lines_.size());
+          lines_.push_back(l);
+        }
+        break;
+      case 2:
+        {
+          MeshObject<2> t;
+          for (unsigned int i=0; i<3; i++)
+            t.nodes[i] = obj_4_node_[mesh_->node_vector.index(edge->side(0)->node(i))];
+          obj_4_edg_.push_back(triangles_.size());
+          triangles_.push_back(t);
+        }
+        break;
+    }
+  }
+}
+
+
+
+void DuplicateNodes::init_from_elements()
+{
+  // initialize the objects from elements
+  FOR_ELEMENTS(mesh_, ele) {
+    switch (ele->dim()) {
+      case 1:
+        {
+          MeshObject<1> l;
+          for (unsigned int i=0; i<2; i++)
+            l.nodes[i] = obj_4_node_[mesh_->node_vector.index(ele->node[i])];
+          for (unsigned int i=0; i<2; i++)
+            l.faces[i] = &points_[obj_4_edg_[ele->side(i)->edge_idx()]];
+          obj_4_el_.push_back(lines_.size());
+          lines_.push_back(l);
+        }
+        break;
+      case 2:
+        {
+          MeshObject<2> tr;
+          for (unsigned int i=0; i<3; i++)
+            tr.nodes[i] = obj_4_node_[mesh_->node_vector.index(ele->node[i])];
+          for (unsigned int i=0; i<3; i++)
+            tr.faces[i] = &lines_[obj_4_edg_[ele->side(i)->edge_idx()]];
+          obj_4_el_.push_back(triangles_.size());
+          triangles_.push_back(tr);
+        }
+        break;
+      case 3:
+        {
+          MeshObject<3> te;
+          for (unsigned int i=0; i<4; i++)
+            te.nodes[i] = obj_4_node_[mesh_->node_vector.index(ele->node[i])];
+          for (unsigned int i=0; i<4; i++)
+            te.faces[i] = &triangles_[obj_4_edg_[ele->side(i)->edge_idx()]];
+          obj_4_el_.push_back(tetras_.size());
+          tetras_.push_back(te);
+        }
+        break;
+    }
+  }
+}
+
+
+/**
+ * The main functionality of the class creates duplicate nodes
+ * at interfaces with fractures.
+ * 
+ * For each node:
+ * 1) Create the lists of elements (node_elements) and edges (node_edges) sharing the node.
+ * 2) Divide node_elements into groups (components) which share one of node_edges.
+ * 3) If more than one component was created, create for each extra component a new node
+ *    and update its index in the elements from this component.
+ *    Also set node_dim_ (dimension of elements using the node).
+ */
+void DuplicateNodes::duplicate_nodes()
+{
+  FOR_NODES(mesh_, n){
+    // create list of elements
```

Comment:
For every node we should create a graph, list of:
```
Vertex is a struct containing {
	uint el_idx, component;
	uint n_neigbours;
	uint neighbours[30];  }
```
We need unordered map: el_idx to vtx_idx
For an mesh edge add the graph edge for every pair of elements connected by the mesh edge.
`vertices[el_to_vtx[el1]].neighbours.push_back(el_to_vtx[el2]`